jdaviz/configs/default/plugins/line_lists/line_lists.py

```diff
@@ -376,8 +376,8 @@ def vue_slider_reset(self, event):
 
     def _on_spectrum_viewer_limits_changed(self, event=None):
         sv = self.app.get_viewer('spectrum-viewer')
-        self.spectrum_viewer_min = sv.state.x_min
-        self.spectrum_viewer_max = sv.state.x_max
+        self.spectrum_viewer_min = float(sv.state.x_min)
+        self.spectrum_viewer_max = float(sv.state.x_max)
```

Comment:
what use-case required casting here?  Can we fix it at the source instead to prevent passing non-floats to the state objects?

Reply:
Trying to load the converted cube into data collection created a long traceback that ended at this point. I can try fixing the original issue again but if I still can't get it I might need some additional help.

Reply:
Here is the traceback if I remove the `float` cast:

<details>

```
---------------------------------------------------------------------------
TraitError                                Traceback (most recent call last)
Input In [2], in <cell line: 2>()
      1 cube = "/Users/javerbukh/Documents/jdaviz_dev/data/data_for_cubeviz/ogle_subcube_1386.fits"
----> 2 viz.load_data(cube)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/core/helpers.py:69, in ConfigHelper.load_data(self, data, parser_reference, **kwargs)
     68 def load_data(self, data, parser_reference=None, **kwargs):
---> 69     self.app.load_data(data, parser_reference=parser_reference, **kwargs)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/app.py:469, in Application.load_data(self, file_obj, parser_reference, **kwargs)
    464         parser = data_parser_registry.members.get(data_parser)
    466 if parser is not None:
    467     # If the parser returns something other than known, assume it's
    468     #  a message we want to make the user aware of.
--> 469     msg = parser(self, file_obj, **kwargs)
    471     if msg is not None:
    472         snackbar_message = SnackbarMessage(
    473             msg, color='error', sender=self)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/cubeviz/plugins/parsers.py:77, in parse_data(app, file_obj, data_type, data_label)
     74                 _parse_esa_s3d(app, hdulist, data_label, ext=ext, viewer_name=viewer_name)
     76         else:
---> 77             _parse_hdulist(app, hdulist, file_name=data_label or file_name)
     79 # If the data types are custom data objects, use explicit parsers. Note
     80 #  that this relies on the glue-astronomy machinery to turn the data object
     81 #  into something glue can understand.
     82 elif isinstance(file_obj, Spectrum1D):

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/cubeviz/plugins/parsers.py:168, in _parse_hdulist(app, hdulist, file_name)
    166 app.add_data_to_viewer('flux-viewer', data_label)
    167 # Add flux to spectrum viewer
--> 168 app.add_data_to_viewer('spectrum-viewer', data_label)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/app.py:912, in Application.add_data_to_viewer(self, viewer_reference, data_path, clear_other_data, ext)
    910 if data_id is not None:
    911     selected_data_items[data_id] = 'visible'
--> 912     self._update_selected_data_items(viewer_item['id'], selected_data_items)
    913 else:
    914     raise ValueError(
    915         f"No data item found with label '{data_label}'. Label must be one "
    916         "of:\n\t" + "\n\t".join([
    917             data_item['name'] for data_item in self.state.data_items]))

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/app.py:1248, in Application._update_selected_data_items(self, viewer_id, selected_items)
   1244 active_data_labels.append(label)
   1246 [data] = [x for x in self.data_collection if x.label == label]
-> 1248 viewer.add_data(data, percentile=95)
   1250 add_data_message = AddDataMessage(data, viewer,
   1251                                   viewer_id=viewer_id,
   1252                                   sender=self)
   1253 self.hub.broadcast(add_data_message)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/specviz/plugins/viewers.py:368, in SpecvizProfileView.add_data(self, data, color, alpha, **layer_state)
    348 """
    349 Overrides the base class to add markers for plotting
    350 uncertainties and data quality flags.
   (...)
    364     `True` if successful, `False` otherwise.
    365 """
    366 # The base class handles the plotting of the main
    367 # trace representing the spectrum itself.
--> 368 result = super().add_data(data, color, alpha, **layer_state)
    370 self._plot_uncertainties()
    372 self._plot_mask()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue_jupyter/view.py:117, in IPyWidgetView.add_data(self, data, color, alpha, **layer_state)
    113 def add_data(self, data, color=None, alpha=None, **layer_state):
    115     data = validate_data_argument(self._data, data)
--> 117     result = super().add_data(data)
    119     if not result:
    120         return

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/common/viewer.py:207, in Viewer.add_data(self, data)
    203     raise IncompatibleDataException("Data not in DataCollection")
    205 # Create layer artist and add to container. First check whether any
    206 # plugins want to make a custom layer artist.
--> 207 layer = get_layer_artist_from_registry(data, self) or self.get_data_layer_artist(data)
    209 if layer is None:
    210     return False

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/common/viewer.py:238, in Viewer.get_data_layer_artist(self, layer, layer_state)
    237 def get_data_layer_artist(self, layer=None, layer_state=None):
--> 238     return self.get_layer_artist(self._data_artist_cls, layer=layer, layer_state=layer_state)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue_jupyter/view.py:142, in IPyWidgetView.get_layer_artist(self, cls, layer, layer_state)
    141 def get_layer_artist(self, cls, layer=None, layer_state=None):
--> 142     return cls(self, self.state, layer=layer, layer_state=layer_state)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue_jupyter/bqplot/profile/layer_artist.py:30, in BqplotProfileLayerArtist.__init__(self, view, viewer_state, layer_state, layer)
     28 def __init__(self, view, viewer_state, layer_state=None, layer=None):
---> 30     super().__init__(viewer_state, layer_state=layer_state, layer=layer)
     32     # Watch for changes in the viewer state which would require the
     33     # layers to be redrawn
     34     self._viewer_state.add_global_callback(self._update_profile)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/common/layer_artist.py:29, in LayerArtist.__init__(self, viewer_state, layer_state, layer)
     25 self.state = layer_state or self._layer_state_cls(viewer_state=viewer_state,
     26                                                   layer=self.layer)
     28 if self.state not in self._viewer_state.layers:
---> 29     self._viewer_state.layers.append(self.state)
     31 self.zorder = self.state.zorder
     32 self.visible = self.state.visible

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/containers.py:52, in CallbackList.append(self, value)
     50 def append(self, value):
     51     super(CallbackList, self).append(self._prepare_add(value))
---> 52     self.notify_all()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/containers.py:45, in CallbackList.notify_all(self, *args, **kwargs)
     43 def notify_all(self, *args, **kwargs):
     44     for callback in self.callbacks:
---> 45         callback(*args, **kwargs)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/containers.py:166, in dynamic_callback.__call__(self, *args, **kwargs)
    165 def __call__(self, *args, **kwargs):
--> 166     self.function(*args, **kwargs)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/containers.py:189, in ListCallbackProperty._default_setter.<locals>.callback(*args, **kwargs)
    188 def callback(*args, **kwargs):
--> 189     self.notify(instance, wrapped_list, wrapped_list)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/core.py:125, in CallbackProperty.notify(self, instance, old, new)
    123     return
    124 for cback in self._callbacks.get(instance, []):
--> 125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
    127     cback(old, new)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/profile/state.py:157, in ProfileViewerState._layers_changed(self, *args)
    155 @defer_draw
    156 def _layers_changed(self, *args):
--> 157     self._update_combo_ref_data()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/profile/state.py:73, in ProfileViewerState._update_combo_ref_data(self)
     72 def _update_combo_ref_data(self):
---> 73     self.ref_data_helper.set_multiple_data(self.layers_data)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/core/data_combo_helper.py:522, in ManualDataComboHelper.set_multiple_data(self, datasets)
    520 for data in unique_data_iter(datasets):
    521     self.append_data(data, refresh=False)
--> 522 self.refresh()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/core/data_combo_helper.py:445, in BaseDataComboHelper.refresh(self, *args)
    444 def refresh(self, *args):
--> 445     self.choices = [data for data in self._datasets]
    446     self.refresh_component_ids()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/core/data_combo_helper.py:86, in ComboHelper.choices(self, choices)
     84 with delay_callback(self.state, self.selection_property):
     85     prop = getattr(type(self.state), self.selection_property)
---> 86     prop.set_choices(self.state, choices)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/core.py:538, in delay_callback.__exit__(self, *args)
    535     self.instance._process_delayed_global_callbacks(resume_props)
    537 for p, args in notifications:
--> 538     p.notify(*args)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/matplotlib/state.py:35, in DeferredDrawSelectionCallbackProperty.notify(self, *args, **kwargs)
     33 @defer_draw
     34 def notify(self, *args, **kwargs):
---> 35     super(DeferredDrawSelectionCallbackProperty, self).notify(*args, **kwargs)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/core.py:127, in CallbackProperty.notify(self, instance, old, new)
    125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
--> 127     cback(old, new)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/profile/state.py:190, in ProfileViewerState._reference_data_changed(self, before, after)
    187                 self.x_att_helper.world_coord = False
    188                 self.x_att = self.reference_data.pixel_component_ids[0]
--> 190         self._update_att()
    192 self.reset_limits()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/profile/state.py:95, in ProfileViewerState._update_att(self, *args)
     93     else:
     94         self.x_att_pixel = self.x_att
---> 95 self._reset_x_limits()

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/profile/state.py:121, in ProfileViewerState._reset_x_limits(self, *event)
    119 with delay_callback(self, 'x_min', 'x_max'):
    120     self.x_min = x_min
--> 121     self.x_max = x_max

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/core.py:538, in delay_callback.__exit__(self, *args)
    535     self.instance._process_delayed_global_callbacks(resume_props)
    537 for p, args in notifications:
--> 538     p.notify(*args)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/utils/matplotlib.py:176, in defer_draw.<locals>.wrapper(*args, **kwargs)
    172 @wraps(func)
    173 def wrapper(*args, **kwargs):
    175     if len(DEFER_DRAW_BACKENDS) == 0:
--> 176         return func(*args, **kwargs)
    178     # Don't recursively defer draws. We just check the first draw_idle
    179     # method since all should be modified in sync.
    180     if isinstance(DEFER_DRAW_BACKENDS[0].draw_idle, DeferredMethod):

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/glue/viewers/matplotlib/state.py:24, in DeferredDrawCallbackProperty.notify(self, *args, **kwargs)
     22 @defer_draw
     23 def notify(self, *args, **kwargs):
---> 24     super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/echo/core.py:125, in CallbackProperty.notify(self, instance, old, new)
    123     return
    124 for cback in self._callbacks.get(instance, []):
--> 125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
    127     cback(old, new)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/default/plugins/line_lists/line_lists.py:111, in LineListTool.__init__.<locals>.<lambda>(x_min)
    106 self.hub.subscribe(self, LineIdentifyMessage,
    107                    handler=self._process_identify_change)
    109 # if set to auto (default), update the slider range when zooming on the spectrum viewer
    110 self._viewer.state.add_callback("x_min",
--> 111                                 lambda x_min: self._on_spectrum_viewer_limits_changed())
    112 self._viewer.state.add_callback("x_max",
    113                                 lambda x_max: self._on_spectrum_viewer_limits_changed())

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/default/plugins/line_lists/line_lists.py:379, in LineListTool._on_spectrum_viewer_limits_changed(self, event)
    377 def _on_spectrum_viewer_limits_changed(self, event=None):
    378     sv = self.app.get_viewer('spectrum-viewer')
--> 379     self.spectrum_viewer_min = sv.state.x_min
    380     self.spectrum_viewer_max = sv.state.x_max
    382     # Also update the slider range

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/traitlets/traitlets.py:715, in TraitType.__set__(self, obj, value)
    713     raise TraitError('The "%s" trait is read-only.' % self.name)
    714 else:
--> 715     self.set(obj, value)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/traitlets/traitlets.py:689, in TraitType.set(self, obj, value)
    688 def set(self, obj, value):
--> 689     new_value = self._validate(obj, value)
    690     try:
    691         old_value = obj._trait_values[self.name]

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/traitlets/traitlets.py:721, in TraitType._validate(self, obj, value)
    719     return value
    720 if hasattr(self, "validate"):
--> 721     value = self.validate(obj, value)  # type:ignore[attr-defined]
    722 if obj._cross_validation_lock is False:
    723     value = self._cross_validate(obj, value)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/traitlets/traitlets.py:2274, in Float.validate(self, obj, value)
   2272     value = float(value)
   2273 if not isinstance(value, float):
-> 2274     self.error(obj, value)
   2275 return _validate_bounds(self, obj, value)

File ~/opt/anaconda3/envs/jdaviz-dev-2.8/lib/python3.8/site-packages/traitlets/traitlets.py:827, in TraitType.error(self, obj, value, error, info)
    821 else:
    822     e = "The '{}' trait expected {}, not {}.".format(
    823         self.name,
    824         self.info(),
    825         describe("the", value),
    826     )
--> 827 raise TraitError(e)

TraitError: The 'spectrum_viewer_min' trait of a LineListTool instance expected a float, not the int64 0.
```

</details>

Reply:
I'm curious if the bump on the version requirement for glue-astronomy removes the need for this?  If not, then let's just leave it in

Reply:
ok, then let's leave this in, it doesn't cost much and shouldn't do any harm.  Thanks for checking!